aws/resource_aws_cognito_user_pool_client.go

```diff
@@ -0,0 +1,220 @@
+			},
+
+			"write_attributes": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"refresh_token_validity": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				Default:      30,
+				ValidateFunc: validateIntegerInRange(0, 3650),
+			},
+		},
+	}
+}
+
+func resourceAwsCognitoUserPoolClientCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cognitoidpconn
+
+	params := &cognitoidentityprovider.CreateUserPoolClientInput{
+		ClientName: aws.String(d.Get("name").(string)),
+		UserPoolId: aws.String(d.Get("user_pool_id").(string)),
+	}
+
+	if v, ok := d.GetOk("generate_secret"); ok {
+		params.GenerateSecret = aws.Bool(v.(bool))
+	}
+
+	if v, ok := d.GetOk("explicit_auth_flows"); ok {
+		params.ExplicitAuthFlows = expandStringList(v.([]interface{}))
+	}
+
+	if v, ok := d.GetOk("read_attributes"); ok {
+		params.ReadAttributes = expandStringList(v.([]interface{}))
+	}
+
+	if v, ok := d.GetOk("write_attributes"); ok {
+		params.WriteAttributes = expandStringList(v.([]interface{}))
+	}
+
+	if v, ok := d.GetOk("refresh_token_validity"); ok {
+		params.RefreshTokenValidity = aws.Int64(int64(v.(int)))
+	}
+
+	log.Printf("[DEBUG] Creating Cognito User Pool Client: %s", params)
+
+	resp, err := conn.CreateUserPoolClient(params)
+
+	if err != nil {
+		return errwrap.Wrapf("Error creating Cognito User Pool Client: {{err}}", err)
+	}
+
+	d.SetId(*resp.UserPoolClient.ClientId)
+	d.Set("user_pool_id", *resp.UserPoolClient.UserPoolId)
+	d.Set("name", *resp.UserPoolClient.ClientName)
+
+	return resourceAwsCognitoUserPoolClientRead(d, meta)
+}
+
+func resourceAwsCognitoUserPoolClientRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cognitoidpconn
+
+	params := &cognitoidentityprovider.DescribeUserPoolClientInput{
+		ClientId:   aws.String(d.Id()),
+		UserPoolId: aws.String(d.Get("user_pool_id").(string)),
+	}
+
+	log.Printf("[DEBUG] Reading Cognito User Pool Client: %s", params)
+
+	resp, err := conn.DescribeUserPoolClient(params)
+
+	if err != nil {
+		if isAWSErr(err, "ResourceNotFoundException", "") {
+			log.Printf("[WARN] Cognito User Pool Client %s is already gone", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	if resp.UserPoolClient.ExplicitAuthFlows != nil {
+		d.Set("explicit_auth_flows", flattenStringList(resp.UserPoolClient.ExplicitAuthFlows))
+	}
+
+	if resp.UserPoolClient.ReadAttributes != nil {
+		d.Set("read_attributes", flattenStringList(resp.UserPoolClient.ReadAttributes))
+	}
+
+	if resp.UserPoolClient.WriteAttributes != nil {
+		d.Set("write_attributes", flattenStringList(resp.UserPoolClient.WriteAttributes))
+	}
+
+	if resp.UserPoolClient.RefreshTokenValidity != nil {
+		d.Set("refresh_token_validity", resp.UserPoolClient.RefreshTokenValidity)
+	}
+	
+	if resp.UserPoolClient.ClientSecret != nil {
+		d.Set("client_secret", resp.UserPoolClient.ClientSecret)
+	}
```

Comment:
`d.Set()` will perform safe dereferencing (and deal with `nil` without crashing), so all of the above 5 checks are redundant. 😉

Reply:
Ah thanks. I've been copying code from elsewhere rather than looking into the intracacies. I'll add that to my changes. Just had to rebuild my dev VM